src/box/lua/config/applier/roles.lua

```diff
@@ -0,0 +1,62 @@
+local log = require('internal.config.utils.log')
+
+local last_loaded = {}
+
+local function stop_roles(roles)
+    for role_name, role in pairs(last_loaded) do
+        if roles == nil or not roles[role_name] then
+            log.verbose('roles.post_apply: stop role ' .. role_name)
+            role.stop()
+        end
+    end
+end
+
+local function apply(config)
+    local configdata = config._configdata
+    local role_names = configdata:get('roles', {use_default = true})
+    if role_names == nil or next(role_names) == nil then
+        stop_roles()
+        return
+    end
+
+    -- Remove duplicates.
+    local roles = {}
+    for _, role_name in pairs(role_names) do
+        roles[role_name] = true
+    end
```

Comment:
I looked on [cartridge's approach](https://github.com/tarantool/cartridge/blob/986e8065ae83747b184fa7655f1f6e415645887b/cartridge/roles.lua#L48) to define the order of loading of the given roles and I liked it.

Previously I said that it seems OK to load roles in arbitrary order, while there are no dependencies between them. I changed my mind. I think that the order should be strictly defined to prevent subtle problems from appear.

My proposal for the resorting stage is the following.

1. Collect all the role names and start to process them one-by-one.
2. `require` the given current role R.
3. If it has no dependencies and was not visited by the algorithm before, add it into resulting list `RS`.
4. For each dependency in the given order check that it is present in the initial list of the roles. If it is not so, give an error. If the check passed, process the dependency role name using this algorith recursively.

Or, the same, as a code.

```lua
#!/usr/bin/env tarantool

local fun = require('fun')
local json = require('json')

-- bar <- foo <- baz
local function requirex(x)
    if x == 'foo' then
        return {
            dependencies = {'baz'},
        }
    elseif x == 'bar' then
        return {
            dependencies = {'foo'},
        }
    elseif x == 'baz' then
        return {}
    else
        assert(false)
    end
end

local function resort_roles(src)
    local resorted = {}

    -- Needed to detect roles that are not listed in the
    -- instance's role list.
    local all_ht = fun.iter(src):map(function(x) return x, true end):tomap()

    -- Needed to detect circular dependencies.
    local to_be_added = {}

    -- To skip already added roles.
    local added_ht = {}

    local function add_role(role_name)
        if added_ht[role_name] then
            return
        end

        local role = requirex(role_name)
        to_be_added[role_name] = true

        for _, dep in ipairs(role.dependencies or {}) do
            -- Detect a role that is not in the list of instance's
            -- roles.
            if not all_ht[dep] then
                error(('Role %q requires role %q, but the latter is not in ' ..
                    'the list of roles of the instance'):format(role_name, dep),
                    0)
            end

            -- Detect a circular depdendency.
            if to_be_added[dep] and role_name == dep then
                error(('Circular dependency: role %q depends on itself'):format(
                    role_name), 0)
            end
            if to_be_added[dep] and role_name ~= dep then
                error(('Circular dependency: roles %q and %q depend on each ' ..
                    'other'):format(role_name, dep), 0)
            end

            -- Go into the recursion: add the dependency.
            add_role(dep)
        end

        -- Add the role itself and track it in the temporary
        -- hashtables.
        to_be_added[role_name] = nil
        added_ht[role_name] = true
        table.insert(resorted, role_name)
    end

    -- Keep the order, where the dependency tree doesn't obligate
    -- us to change it.
    for _, role_name in ipairs(src) do
        add_role(role_name)
    end

    return resorted
end

print(json.encode(resort_roles({'foo', 'bar', 'baz', 'foo'})))
-- ["bar", "foo", "bar"]
```

While we have no dependencies, I think, it is OK to guarantee the order of `require`/`validate`/`apply` is the same as written in the config (except excluding duplicates).

Reply:
And, while we're here, I would (like cartridge) define some specific order between `apply` and `stop`.

For example, we can perform all stops before all starts. The stops should also be performed in some guaranteed order. I propose the reverse of the starting order (reverse of `resort_order`'s result).

As another example, cartridge forms a list of all known roles (from previous and from the next configuration) and perform start/stop in the order of the list (depending on whether the role is present in the new config). I think it is error-prone to stop dependencies before a role that depends on them, but maybe there is something useful in the overall approach.

Anyway, while we have no dependencies, we still should define the order between starts and stops. I propose to do all the stops in the reversed order and then all starts in the direct order.

Reply:
Since there is no dependency, I used a slightly simpler approach. In addition, roles are now loaded, validated, and applied in the same order in which they were given. Stopped in reversed order.